### PR TITLE
Add OceanBC interface modeled on AtmosBC interface

### PR DESCRIPTION
**Pull request by blallen**
# Description

Updating the Ocean boundary conditions interface to match that used in Atmos. Makes specifying boundary conditions easier for users and should make the move to a cubed sphere easier. 

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by blallen** [comment]
@simonbyrne here's my first pass at matching ocean boundary conditions to your interface. Everything runs right now, but the boundary conditions aren't being applied correctly yet. I'll be working on that throughout the week

@jkozdon and @sandreza, can you check the `bc_temperature.jl` and `bc_velocity.jl` files to make sure I'm actually applying everything correctly and in correspondence with the design docs?

**Comment by blallen** [comment]
Just rebased on top of master and got this fresh again.

Three big improvements beyond that:

1) implement Neumann boundary conditions only using boundary states. @simonbyrne take a look at this and let me know if it will also work for the atmosphere. Having one place to implement all BCs should be easier for developers.

2) Realized that the ocean surface can't be `Impenetrable` in a hydrostatic boussinesq set up, so I implemented a `Penetrable` mass/velocity boundary condition with free slip and kinematic stress options.

3) Started working to find a nice interface for specifying different physical boundary fluxes on different boundaries. This part is still WIP and can be factored out into a separate PR.

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
@jm-c @christophernhill, here's the new interface for specifying boundary conditions for the ocean. Things are split out into velocity and temperature boundary conditions, so you can mix and match between the two. Also, the boundary conditions should apply in any geometry, not just the box (assuming the `kinematic_stress` and `surface_flux` functions are general enough). 

@simonbyrne for this round, I'm not letting users pass functions to the flux boundary conditions, so each boundary will get the same flux forcings. I have some WIP in progress on passing a function like you do with AtmosBC, but that will come in a separate PR. I want to get this stuff in because it's been working for a while.

also @simonbyrne, take a look at how I did the flux boundary conditions and let me know if you think this makes sense for the Atmos model as well. @charleskawczynski and @gagnelandmanna have already implemented similar approaches in the soil models.

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try-

**Comment by blallen** [comment]
bors try

File: src/Ocean/HydrostaticBoussinesq/BoundaryConditions.jl [hunk 1]
```diff
@@ -3,6 +3,20 @@ using ..DGmethods.NumericalFluxes:
...
+            temperature = Insulating())
+
+The standard boundary condition for OceanModel. The default options imply a "no flux" boundary condition.
+"""
+
+Base.@kwdef struct OceanBC{M, T}
+    velocity::M = Impenetrable(NoSlip())
+    temperature::T = Insulating()
+end
+
```
Comment (charleskawczynski) [inline src/Ocean/HydrostaticBoussinesq/BoundaryConditions.jl:19]:
```suggestion
```
We can’t have new lines between doc strings and structs/functions

File: src/Ocean/HydrostaticBoussinesq/BoundaryConditions.jl [hunk 2]
```diff
@@ -3,6 +3,20 @@ using ..DGmethods.NumericalFluxes:
...
 
+export OceanBC
+
+"""
+    OceanBC(velocity    = Impenetrable(NoSlip())
+            temperature = Insulating())
+
+The standard boundary condition for OceanModel. The default options imply a "no flux" boundary condition.
+"""
+
```
Comment (charleskawczynski) [inline src/Ocean/HydrostaticBoussinesq/BoundaryConditions.jl:14]:
```suggestion
```
We can’t have new lines between doc strings and structs/functions

File: src/Ocean/HydrostaticBoussinesq/bc_velocity.jl [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+"""
+struct NoSlip <: VelocityDragBC end
+
+"""
+    ocean_velocity_boundary_state!(::NumericalFluxFirstOrder, ::Impenetrable{NoSlip}, ::HBModel)
+
+apply no slip boundary condition for velocity
+sets reflective ghost point
+"""
+
```
Comment (charleskawczynski) [inline src/Ocean/HydrostaticBoussinesq/bc_velocity.jl:41]:
```suggestion
```

**Comment by jkozdon** [comment]
> No opinion on this one.

But seems like a good idea to have a unified interface. Thanks!

**Comment by blallen** [comment]
Bors r+

**Comment by blallen** [comment]
Just to note, this PR fixes a mistake introduced in PR #660. When that PR switched us from a central boundary flux formulation 

https://github.com/CliMA/ClimateMachine.jl/blob/461657afa53ad608015a27f7c91dd8c15e605810/src/DGmethods/NumericalFluxes.jl#L244-L268

to an inflow boundary flux formulation

https://github.com/CliMA/ClimateMachine.jl/blob/d808788447a2354d1e5927b986d212780f078278/src/DGmethods/NumericalFluxes.jl#L241-L263

for the boundary flux for the second-order/diffusive fluxes, we should have changed 

https://github.com/CliMA/ClimateMachine.jl/blob/c9bc1b6598d1340b238f5c22b85ae745f8954ee5/src/Ocean/Model/HydrostaticBoussinesqModel.jl#L463
and
https://github.com/CliMA/ClimateMachine.jl/blob/c9bc1b6598d1340b238f5c22b85ae745f8954ee5/src/Ocean/Model/HydrostaticBoussinesqModel.jl#L470

to `D⁺.∇u = Diagonal(A⁺.ν) \  τ` and `D⁺.∇θ = Diagonal(A⁺.κ) \  σ` to match. (The left divides got reverted in a PR #841 when we moved the viscosity and diffusivities back into the gradient fluxes). 

This omission caused the code to apply double the prescribed wind stress and temperature forcing during the period between PR #600 (Feb 5 2020) and PR #851 (May 30 2020). When updating the boundary condition treatment for this PR, I corrected the mistake.

**Pull request merged**